serial/serial_resources/component_instance.py

```diff
@@ -0,0 +1,112 @@
+import serial
+from ..api_client import APIClient
+from .component_instance_link import ComponentInstanceLink
+
+class ComponentInstance:
+    def __init__(self, component_instance_data):
+        self.client = APIClient(serial.api_key, serial.base_url)
+        self.data = component_instance_data
+        self.links = [] 
+
+    def add_link(self, link_name, child_identifier, break_prior_links=False, process_entry=None):
+        """
+        Creates a link between this component instance and a child at a specific
+        process_entry
+
+        Args: 
+        - link_name: User facing name for the link
+        - child_identifier: Identifier for the child component instance to be linked
+        - break_prior_links: Boolean for whether to break prior links
+        - process_entry?: Override for the process entry this component instance object 
+        was created with. Must be included if this component instance object was not 
+        created with a process entry.
+
+        Returns:
+        - New component instance link
+        """
+
+        if serial.debug:
+            print(f"Adding link: {link_name} with child identifier: {child_identifier}")
+        child_component_instance_params = {
+                "identifier": child_identifier
+                }
+        link_params = {
+                "name": link_name
+                }
+        child_component_instance = self.client.make_api_request("/components/instances",
+                                                                "GET",
+                                                                params=child_component_instance_params)[0]
+        link_dataset = self.client.make_api_request("/datasets",
+                                                    "GET",
+                                                    params=link_params)[0]
+        process_entry_id = None
+        if process_entry:
+            process_entry_id = process_entry.id
+        else:
+            process_entry_id = self.process_entry.id
+        data = {
+                "parent_component_instance_id": self.data["id"],
+                "child_component_instance_id": child_component_instance["id"],
+                "dataset_id": link_dataset["dataset"]["id"],
+                "process_entry_id": process_entry_id,
+                "break_prior_links": break_prior_links,
+                }
+        self.links.append(ComponentInstanceLink(self.client.make_api_request("/components/instances/links", "PUT", data=data)))
+        return self.links[-1]
+
+class ComponentInstances:
```

Comment:
Docstring needed for these data object classes (I should also remove the `__init__` from here since they're never meant to be instantiated on their own.